ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,358 @@
+
+    {connect_timeout, tcp_opts} = Keyword.pop(tcp_opts, :connect_timeout, @default_timeout)
+
+    state = %State{
+      host: to_host(host),
+      port: port,
+      tcp_opts: tcp_opts,
+      loqui_path: loqui_path,
+      connect_timeout: connect_timeout,
+      sequence: 1,
+      registered_compressors: registered_compressors,
+      registered_codecs: registered_codecs,
+      buffer: <<>>}
+
+    {:connect, :init, state}
+  end
+
+  def close(conn, timeout \\ 5000),
+    do: Connection.call(conn, :close, timeout)
+
+  def ping(conn, timeout \\ 5000) do
+    Connection.call(conn, :ping, timeout)
+  end
+
+  def request(conn, payload, timeout \\ 5000) do
+    Connection.call(conn, {:request, payload}, timeout)
+  end
+
+  def push(conn, payload) do
+    Connection.cast(conn, {:push, payload})
+  end
+
+  def connect(_info, %{sock: nil, host: host, port: port, tcp_opts: tcp_opts, connect_timeout: connect_timeout}=state) do
+    with {:ok, sock}  <- :gen_tcp.connect(host, port, tcp_opts, connect_timeout),
+         {:ok, state} <- do_upgrade(%{state | sock: sock}),
+         state        <- make_active_once(state),
+         {:ok, state} <- do_loqui_connect(state) do
+
+      {:ok, state}
+    else
+      {:error, _} = error ->
+        Logger.error("Couldn't connect to #{host}:#{port} because #{inspect error}")
+        {:stop, error, state}
+    end
+  end
+
+  def disconnect(info, %State{sock: sock}=state) do
+    :ok = :gen_tcp.close(sock)
+    case info do
+      {:close, from} ->
+        Connection.reply(from, :ok)
+        {:stop, :normal, %{state | sock: nil}}
+
+      {:error, reason}  ->
+        {:stop, reason, %{state | sock: nil}}
+    end
+  end
+
+  # Connection callbacks
+  def handle_call(:ping, caller, %{sock: sock}=state) do
+    {next_seq, state} = next_sequence(state)
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_call(_, _from, %State{sequence: :go_away}=state) do
+    {:reply, {:error, :remote_went_away}, state}
+  end
+
+  def handle_call(:close, from, %State{sock: sock}=s) do
+    go_away_packet = Frames.goaway(0, 0, "Closing")
+    :gen_tcp.send(sock, go_away_packet)
+
+    {:disconnect, {:close, from}, s}
+  end
+
+  def handle_call({:request, payload}, caller, %State{sock: sock, codec: codec}=state) do
+    {next_seq, state} = next_sequence(state)
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.request(0, next_seq, encoded_payload))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_cast(_, %State{sequence: :go_away}=state),
+    do: {:noreply, state}
+
+  def handle_cast({:push, payload}, %State{sock: sock, codec: codec}=state) do
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.push(0, encoded_payload))
```

Comment:
Probably send to `self()` after send, and accumulate a buffer until that message arrives, otherwise will delay sending. Hopefully everything can be accumulated without getting scheduled out too many times. You could limit the buffer size before the next send.